### PR TITLE
8274735: javax.imageio.IIOException: Unsupported Image Type  while processing a valid JPEG image

### PR DESCRIPTION
**Pull request by prrace**
The JDK's built-in Image I/O JPEG plugin decodes JPEGs which are YCbCr and RGb and Grayscale but rejects CMYK and YCCK JPEGs.
These would correspond to JFIF encapsulated JPEGs.

This was a reasonable decision to limit support to what was actually likely to be seen - CMYK JPEGs probably should just be created for direct to print devices and never seen on the internet as an interchange format.

Indeed this (below) RFC for mail suggests JFIF is what clients should transmit :

https://datatracker.ietf.org/doc/html/rfc2046 :  Multipurpose Internet Mail Extensions - Part 2

   A media type of "image" indicates that the body contains an image.
   The subtype names the specific image format.  These names are not
   case sensitive. An initial subtype is "jpeg" for the JPEG format
   using JFIF encoding [JPEG].

And for many years, browsers either failed to display CMYK JPEGs or displayed them incorrectly.

But alas they do crop up and when they do, Image I/O rejects them - to the surprise and
consternation of some developer who knows little of JPEGs and arcane colour spaces.

Full support of CMYK JPEGs via the whole Image I/O API including proper handling as metadata
will lead us down a path that might even extend into EXIF support for JPEG and likely will
require the JPEG Metadata spec to be updated. 

So the goal here is to skirt that as much as possible by just 
(1) not rejecting read of CMYK JPEGs and returning a BufferedImage with a suitable ColorModel
(2) allowing the same BufferedImage with that ColorSpace to be written back out again as CMYK JPEGs 

Reading will preserve any embedded ICC Color Profile, but writing will not re-encode it since
that requires work on metadata. That will have to be a follow-up fix, but at the very least it
means that a re-written JPEG will benefit from a default treatment CMYK by any other client.

A test is provided which verifies that the colours are interpreted reasonably, both on initial read
and on write and re-read. 

JPEGs with embedded profiles have been tested on read, but since there are licensing issues with those,
the test uses JPEGs which have no embedded profile and depend on an internal implementation class.

That class has its hashCode() method updated because testing showed that toString() invoked hashCode()
in a way that caused a StackOverflowError - so nothing per se to do with this fix but a necessary update.

Anyone who has trustworthy CMYK JPEGs that render properly in (eg) major web browsers but not with this
code can submit those JPEGs so I can evaluate.

But all our automated tests pass as well as some manual testing and evaluation etc.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8274735](https://bugs.openjdk.java.net/browse/JDK-8274735): javax.imageio.IIOException: Unsupported Image Type  while processing a valid JPEG image

### Reviewers
 * [Alexander Zuev](https://openjdk.java.net/census#kizune) (@azuev-java - **Reviewer**)
 * [Sergey Bylokhov](https://openjdk.java.net/census#serb) (@mrserb - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7849/head:pull/7849` \
`$ git checkout pull/7849`

Update a local copy of the PR: \
`$ git checkout pull/7849` \
`$ git pull https://git.openjdk.java.net/jdk pull/7849/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7849`

View PR using the GUI difftool: \
`$ git pr show -t 7849`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7849.diff">https://git.openjdk.java.net/jdk/pull/7849.diff</a>

</details>

### CODE REVIEW

**Comment by mrserb** [comment]
>Integration blocker
> ⚠️ Failed to retrieve information on issue 8274735. Please make sure it exists and is accessible.

Looks like the JDK-8274735 is closed.

File: src/java.desktop/share/classes/com/sun/imageio/plugins/common/SimpleCMYKColorSpace.java [hunk 1]
```diff
@@ -63,7 +63,7 @@ public boolean equals(Object o) {
     }
 
     public int hashCode() {
-        return theInstance.hashCode();
+        return getClass().hashCode()+7;
```
Comment (ExE-Boss) [inline src/java.desktop/share/classes/com/sun/imageio/plugins/common/SimpleCMYKColorSpace.java:66]: Alternatively, you could do:
```suggestion
        return System.identityHashCode(theInstance);
```
which seems to have been the original intent.

**Review by ExE-Boss — COMMENTED** [review]
Note that [`gray_cmyk.jpg`] ([![gray_cmyk.jpg]][`gray_cmyk.jpg`]) looks more like a greyish brown (`#4A4642`) on my end.

[`gray_cmyk.jpg`]: https://github.com/openjdk/jdk/blob/2573ebdc1afe0c46b621c39195ffd49bf2f724a2/test/jdk/javax/imageio/plugins/jpeg/CMYK/gray_cmyk.jpg
[gray_cmyk.jpg]: https://raw.githubusercontent.com/openjdk/jdk/2573ebdc1afe0c46b621c39195ffd49bf2f724a2/test/jdk/javax/imageio/plugins/jpeg/CMYK/gray_cmyk.jpg

File: src/java.desktop/share/classes/com/sun/imageio/plugins/common/SimpleCMYKColorSpace.java (hunk 1 above)
Comment (prrace) [reply src/java.desktop/share/classes/com/sun/imageio/plugins/common/SimpleCMYKColorSpace.java:66]: I can do that.

**Comment by prrace** [comment]
> Note that [`gray_cmyk.jpg`](https://github.com/openjdk/jdk/blob/2573ebdc1afe0c46b621c39195ffd49bf2f724a2/test/jdk/javax/imageio/plugins/jpeg/CMYK/gray_cmyk.jpg) ([](https://github.com/openjdk/jdk/blob/2573ebdc1afe0c46b621c39195ffd49bf2f724a2/test/jdk/javax/imageio/plugins/jpeg/CMYK/gray_cmyk.jpg)) looks more like a greyish brown (`#4A4642`) on my end.

It was created using java.awt.Color.gray which is 128.128,128 - ie #808080 as RGB, saved as JPG then converted via an online tool to CMYK  and the round trip tolerance check is passing when comparing to the original gray color, so it must be something about whatever you are using to display it and how it interprets that.

File: src/java.desktop/share/classes/com/sun/imageio/plugins/jpeg/JPEGImageWriter.java [hunk 2]
```diff
@@ -1721,6 +1736,12 @@ private void grabPixels(int y) {
                                         srcBands);
         }
         raster.setRect(sourceLine);
+        if (invertCMYK) {
+            byte[] data = ((DataBufferByte)raster.getDataBuffer()).getData();
```
[the remark this answers is not in the document]
Comment (prrace) [reply src/java.desktop/share/classes/com/sun/imageio/plugins/jpeg/JPEGImageWriter.java:1740]: yes

File: src/java.desktop/share/classes/com/sun/imageio/plugins/common/SimpleCMYKColorSpace.java [hunk 3]
```diff
@@ -63,7 +63,7 @@ public boolean equals(Object o) {
     }
 
     public int hashCode() {
-        return theInstance.hashCode();
+        return System.identityHashCode(theInstance);
```
[the remark this answers is not in the document]
Comment (prrace) [reply src/java.desktop/share/classes/com/sun/imageio/plugins/common/SimpleCMYKColorSpace.java:66]: If toString() is called  then you get StackOverflowException(). 
I am surprised no-one hit this already but I hit it repeatedly during testing.

File: src/java.desktop/share/classes/com/sun/imageio/plugins/jpeg/JPEGImageWriter.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (mrserb) [reply src/java.desktop/share/classes/com/sun/imageio/plugins/jpeg/JPEGImageWriter.java:1740]: As far as I understand the in the read code we create a DataBufferByte so later we can use the cast. But in this method we do not create it, and use what the user provide to us, can the user "craft custom " raster/metadata and the cs type?

**Review by azuev-java — APPROVED** [review]
Tested it with some of the CMYK files that i created - seems like all the color conversions are valid and comparing screenshot of the decoded image with the native applications that support the format like Preview or Adobe Photoshop i see no noticeable deviations.

File: src/java.desktop/share/classes/com/sun/imageio/plugins/jpeg/JPEGImageWriter.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (prrace) [reply src/java.desktop/share/classes/com/sun/imageio/plugins/jpeg/JPEGImageWriter.java:1740]: Yes we do. This raster is created here in the ImageWriter as follows 

        // Create an appropriate 1-line databuffer for writing
        int lineSize = sourceWidth*numBandsUsed;

        DataBufferByte buffer = new DataBufferByte(lineSize);

        // Create a raster from that
        int [] bandOffs = JPEG.bandOffsets[numBandsUsed-1];

        raster = Raster.createInterleavedRaster(buffer,
                                                sourceWidth, 1,
                                                lineSize,
                                                numBandsUsed,
                                                bandOffs,
                                                null);

**Review by mrserb — APPROVED** [review]

**Pull request closed without merging**